clouddrift/sphere.py

```diff
@@ -435,3 +435,133 @@ def sphere_to_plane(
+    -------
+    x : np.ndarray
+        x-coordinates in 3D in meters.
+    y : np.ndarray
+        y-coordinates in 3D in meters.
+    z : np.ndarray
+        z-coordinates in 3D in meters.
+
+    Examples
+    --------
+    >>> earthsphere_to_cartesian(np.array([0, 45]),np.array([0, 45]))
+    (array([6378100., 3189050.]),
+    array([      0., 3189050.]),
+    array([      0.        , 4509997.76108592]))
+
+    >>> earthsphere_to_cartesian(np.array([0, 45, 90]),np.array([0, 90, 180]),earth_radius=1)
+    (array([ 1.00000000e+00,  4.32978028e-17, -6.12323400e-17]),
+    array([ 0.00000000e+00,  4.32978028e-17, -1.00000000e+00]),
+    array([0.0000000e+00, 1.0000000e+00, 1.2246468e-16]))
+
+    >>> x, y, z = earthsphere_to_cartesian(np.array([0,5]),np.array([0,5]))
+
+    Raises
+    ------
+    AttributeError
+        If ``lon`` and ``lat`` are not NumPy arrays.
+
+    See Also
+    --------
+    :func:`cartesian_to_earthsphere`
+    """
+
+    x = earth_radius * np.cos(lat * np.pi / 180) * np.cos(lon * np.pi / 180)
+    y = earth_radius * np.cos(lat * np.pi / 180) * np.sin(lon * np.pi / 180)
+    z = earth_radius * np.sin(lat * np.pi / 180)
+
+    x = np.nan_to_num(x, posinf=np.nan, neginf=np.nan)
+    y = np.nan_to_num(y, posinf=np.nan, neginf=np.nan)
+    z = np.nan_to_num(z, posinf=np.nan, neginf=np.nan)
+
+    return x, y, z
+
+
+def cartesian_to_earthsphere(
+    x: np.ndarray,
+    y: np.ndarray,
+    z: np.ndarray,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Converts Cartesian threedimensional coordinates to latitude and longitude on the
+    spherical Earth.
+
+    The Cartesian coordinate system is a right-handed system whose
+    origin lies at the center of the sphere.  It is oriented with the
+    Z-axis passing through the poles and the X-axis passing through
+    the point lon = 0, lat = 0. This function is inverted by `earthsphere_to_cartesian`.
+
+    Parameters
+    -------
+    x : np.ndarray
+        x-coordinates in 3D.
+    y : np.ndarray
+        y-coordinates in 3D.
+    z : np.ndarray
+        z-coordinates in 3D.
+
+    Returns
+    ----------
+    lon : np.ndarray
+        An N-d array of longitudes in degrees in range [-180, 180].
+    lat : np.ndarray
+        An N-d array of latitudes in degrees.
+
+    Examples
+    --------
+    >>> x = EARTH_RADIUS_METERS * np.cos(45 * np.pi / 180)
+    >>> y = EARTH_RADIUS_METERS * np.cos(45 * np.pi / 180)
+    >>> z = 0 * x
+    >>> cartesian_to_earthsphere(x, y, z)
+    (44.99999999999985, 0.0)
+
+    `cartesian_to_earthsphere` is inverted by `earthsphere_to_cartesian`:
+
+    >>> x, y, z = earthsphere_to_cartesian(np.array([45]),np.array(0))
+    >>> cartesian_to_earthsphere(x, y, z)
+    (array([45.]), array([0.]))
+
+    Raises
+    ------
+    AttributeError
+        If ``x``, ``y``, and ``z`` are not NumPy arrays.
+
+    See Also
+    --------
+    :func:`earthsphere_to_cartesian`
+    """
+
+    R = np.sqrt(np.abs(x) ** 2 + np.abs(y) ** 2 + np.abs(z) ** 2)
```

Comment:
Are `abs()` needed? It seems to me they make a difference only when inputs are complex; is that expected?